pkg/user/persister/persister.go

```diff
@@ -213,42 +213,58 @@ func (p *persister) runWorker() error {
 }
 
 func (p *persister) handleChunk(chunk map[string]*puller.Message) {
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+	defer cancel()
+	events := make(map[string][]*eventproto.UserEvent)
+	messages := make(map[string][]*puller.Message)
 	for _, msg := range chunk {
-		event, err := p.unmarshalMessage(msg)
-		if err != nil {
-			handledCounter.WithLabelValues(codes.BadMessage.String()).Inc()
+		// Extract the user event
+		event := p.extractUserEvent(msg)
+		if event == nil {
 			continue
 		}
-		if !p.validateEvent(event) {
-			handledCounter.WithLabelValues(codes.BadMessage.String()).Inc()
-			continue
+		// Append events per environment
+		listEvents, ok := events[event.EnvironmentNamespace]
```

Comment:
The upsert function is per environment. So here we split the chunk messages per environment.